### PR TITLE
Debugging in docker

### PR DESCRIPTION
**Pull request by gpbowman-git**
[API-2966](https://vajira.max.gov/browse/API-2966) - Setting up Vets-api on Windows running in docker.

See the readme - [debugging_in_docker.md](https://github.com/department-of-veterans-affairs/vets-api/blob/debugging_in_docker/docs/setup/debugging_in_docker.md) for instructions on getting vets-api configured to run in a docker container with the RubyMine debugger listening for breakpoints.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -48,7 +48,7 @@ RUN curl -sSL -o /usr/local/bin/cc-test-reporter https://codeclimate.com/downloa
 COPY --chown=vets-api:vets-api config/freshclam.conf docker-entrypoint.sh ./
 USER vets-api
 # XXX: this is tacky
-RUN freshclam --config-file freshclam.conf
+#RUN freshclam --config-file freshclam.conf todo uncomment this before merge
```
Comment (dbrown-va) [inline Dockerfile:51]: Have a `todo` left here to revert this comment.

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -1,7 +1,16 @@
 #!/bin/bash -e
 
-# note this logic is duplciated in the Dockerfile for prod builds,
+# note this logic is duplicated in the Dockerfile for prod builds,
 # if you make major alteration here, please check that usage as well
+whoami
+ls -la /srv/vets-api/src/
+ls -la /srv/vets-api/src/Gemfile.lock
```
Comment (dbrown-va) [inline docker-entrypoint.sh:7]: These debug calls can be removed now that permissions were sorted out.

File: Rakefile [hunk 3]
```diff
@@ -4,10 +4,10 @@
 # for example lib/tasks/capistrano.rake, and they will automatically be available to Rake.
 
 require_relative 'config/application'
-
+# Rake.add_rakelib('./modules/va_forms/lib/tasks')
+# Rake.load_rakefile('./modules/va_forms/lib/tasks/va_forms.rake')
```
Comment (dbrown-va) [inline Rakefile:8]: Couple unused lines that can be removed.

**Review by charleystran — CHANGES_REQUESTED** [review]
Can you revert the Gemfile changes that remove Sidekick enterprise?

**Comment by cshupp1** [comment]
> Can you revert the Gemfile changes that remove Sidekick enterprise?



> Can you revert the Gemfile changes that remove Sidekick enterprise?

@charleystran is everything OK now? I removed those changes.

File: rakelib/docker_debugging.rake [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        Rake::Task['db:create'].invoke
+        Rake::Task['db:schema:load'].invoke
+        Rails.env = 'test'
+        Rake::Task['db:schema:load'].invoke
+        puts 'All dun!'
+      end
+      sh 'foreman start -m all=1,clamd=0,freshclam=0'
+    end
+  end
+end
```
Comment (f1337) [inline rakelib/docker_debugging.rake:28]: Why do we need a new rake task for this? What are we trying to solve?

**Review by f1337 — CHANGES_REQUESTED** [review]
Thank you! I don't hate it at all, but I do have questions & suggestions.

File: docs/setup/debugging_in_docker.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
+### Debugging in Docker, useful for windows developers.
```
[the remark this answers is not in the document]
Comment (f1337) [reply docs/setup/debugging_in_docker.md:1]: Could we also add at the top some kind of disclaimer like: "UNSUPPORTED. These are community-generated steps. YMMV."

File: Gemfile [hunk 6]
```diff
@@ -125,6 +125,8 @@ group :development do
   # POSIX systems should have this already, so we're not going to bring it in on other platforms
   gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
   # Access an IRB console on exception pages or by using <%= console %> in views
+  gem 'debase'
+  gem 'ruby-debug-ide', git: 'https://github.com/corgibytes/ruby-debug-ide', branch: 'feature-add-fixed-port-range'
```
[the remark this answers is not in the document]
Comment (cshupp1) [reply Gemfile:129]: This is the issue:
https://github.com/ruby-debug/ruby-debug-ide/issues/107
Raised in 2017, the fix still isn't merged into main yet.

File: docker-compose.yml [hunk 7]
```diff
@@ -27,6 +27,9 @@ services:
       - dev_bundle:/usr/local/bundle
     ports:
       - "3000:3000"
+      - "26162:26162"
+      - "1234:1234"
+      - "58430-58450:58430-58450"
```
[the remark this answers is not in the document]
Comment (cshupp1) [reply docker-compose.yml:32]: Will do

File: modules/va_forms/app/workers/va_forms/form_reloader.rb [hunk 8]
```diff
@@ -45,7 +45,8 @@ def faraday_options
           verify: false
         }
       }
-      options[:proxy] = { uri: URI.parse('socks://localhost:2001') } unless Rails.env.production?
+      socks = Settings.docker_debugging&.socks_url ? Settings.docker_debugging.socks_url : 'socks://localhost:2001'
```
[the remark this answers is not in the document]
Comment (cshupp1) [reply modules/va_forms/app/workers/va_forms/form_reloader.rb:48]: We can, but this is why I did it this why, consider this portion of my local settings:
```
docker_debugging:
  set_up_db: false
  hang_around: false
  #socks_url: socks://localhost:2001
  #socks_url: socks://172.17.0.1:2001
  #above should work when the host machine is linux
  socks_url: socks5h://host.docker.internal:2001 
```
The original line always assumed the socks proxy is running on localhost.  It is for us too but when running in docker I have to give it the clue 'host.docker.internal' meaning the 'real' localhost.  I personally like it in docker debugging as it forces the key lumped there.  That being said, I am currently incapable of imagining another use case for making it more generic.

File: rakelib/docker_debugging.rake (hunk 4 above)
Comment (cshupp1) [reply rakelib/docker_debugging.rake:28]: From docker-entrypoint.sh:
```
if [ -e  "./docker_debugging" ] ; then
  echo starting rake docker_debugging:setup
  rake docker_debugging:setup
fi
```

After a conversation with Bastos he gave the setup he uses to set up the DB.  
From the local settings (above) I drive the behaviour docker-compose up does for me.

It sets up my DB (if that flag is set) then starts the rails server.  Usually, the DB flag is false so between 'ups' I persist my data.  There is also the 'hang_around' flag.  This ensures that vets_api container doesn't die allowing us to exec in and do whatever we want before any of the various services are started (it is our responsibility to start the rails server if desired).

File: docs/setup/debugging_in_docker.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (f1337) [reply docs/setup/debugging_in_docker.md:1]: `debugging_with_docker_rubymine_windows.md`

[the remark this answers is not in the document]
Comment (cshupp1) [reply docs/setup/debugging_in_docker.md]: Added

**Review by f1337 — APPROVED** [review]
Thank you for the submission and the edits @cshupp1! :thumbsup:
Shaun The Sheep giving thumbs up and smiling

**Review by charleystran — APPROVED** [review]

**Pull request merged**